Core/HLE/sceNetAdhoc.cpp

```diff
@@ -1945,15 +2263,21 @@ static int sceNetAdhocPtpConnect(int id, int timeout, int flag) {
 				
 				// Grab Peer IP
 				if (resolveMAC(&socket->paddr, (uint32_t *)&sin.sin_addr.s_addr)) {
+					// Some games (ie. PSP2) might try to talk to it's self, not sure if they talked through WAN or LAN when using public Adhoc Server tho
+					sin.sin_port = htons(socket->pport + ((isOriPort && !isPrivateIP(sin.sin_addr.s_addr)) ? 0 : portOffset));
+
 					// Grab Nonblocking Flag
 					uint32_t nbio = getBlockingFlag(socket->id);
-					// Switch to Nonblocking Behaviour
-					if (nbio == 0) {
+
+					// Switch to Nonblocking Behaviour. Forcing blocking behaviour on the first connect may fix connection issue on GvG Next Plus, But i don't like using blocking socket with infinite timeout if the game it self were asking for non-blocking behaviour :(
+					/*if (nbio == 0) 
+					{
 						// Overwrite Socket Option
 						changeBlockingMode(socket->id, 1);
-					}
+					}*/
 					
 					// Connect Socket to Peer (Nonblocking)
+					// NOTE: Based on what i read at stackoverflow, The First Non-blocking POSIX connect will always returns EAGAIN/EWOULDBLOCK because it returns without waiting for ACK/handshake, But GvG Next Plus is treating non-blocking PtpConnect just like blocking connect, May be on a real PSP the first non-blocking sceNetAdhocPtpConnect can be successfull?
```

Comment:
@adenovan  you might want to confirm this on real PSP.
Based on how GvG Next Plus is using non-blocking PtpConnect like a blocking one making me to have suspicion, that a non-blocking PtpConnect might have the possibility to be successfully connected on the first attempt.

While a non-blocking POSIX connect are not possible to be successful on the first attempt due to returning without waiting for ACK (at least based on what i read at stackoverflow, never tested this my self)